src/hotspot/cpu/s390/macroAssembler_s390.cpp

```diff
@@ -3134,18 +3109,177 @@ void MacroAssembler::check_klass_subtype_slow_path(Register Rsubklass,
+  }
   final_jmp(*L_success);
 
   // Exit to the surrounding code.
   BIND(L_fallthrough);
-#undef local_brc
 #undef final_jmp
+  BLOCK_COMMENT("} check_klass_subtype_slow_path_linear");
+}
+
+// If Register r is invalid, remove a new register from
+// available_regs, and add new register to regs_to_push.
+Register MacroAssembler::allocate_if_noreg(Register r,
+                                           RegSetIterator<Register> &available_regs,
+                                           RegSet &regs_to_push) {
+  if (!r->is_valid()) {
+    r = *available_regs++;
+    regs_to_push += r;
+  }
+  return r;
+}
+
+// check_klass_subtype_slow_path_table() looks for super_klass in the
+// hash table belonging to super_klass, branching to L_success or
+// L_failure as appropriate. This is essentially a shim which
+// allocates registers as necessary and then calls
+// lookup_secondary_supers_table() to do the work. Any of the temp
+// regs may be noreg, in which case this logic will choose some
+// registers push and pop them from the stack.
+void MacroAssembler::check_klass_subtype_slow_path_table(Register sub_klass,
+                                                         Register super_klass,
+                                                         Register temp_reg,
+                                                         Register temp2_reg,
+                                                         Register temp3_reg,
+                                                         Register temp4_reg,
+                                                         Register result_reg,
+                                                         Label* L_success,
+                                                         Label* L_failure,
+                                                         bool set_cond_codes) {
+  BLOCK_COMMENT("check_klass_subtype_slow_path_table {");
+
+  RegSet temps = RegSet::of(temp_reg, temp2_reg, temp3_reg, temp4_reg);
+
+  assert_different_registers(sub_klass, super_klass, temp_reg, temp2_reg, temp4_reg);
+
+  Label L_fallthrough;
+  int label_nulls = 0;
+  if (L_success == nullptr)   { L_success   = &L_fallthrough; label_nulls++; }
+  if (L_failure == nullptr)   { L_failure   = &L_fallthrough; label_nulls++; }
+  assert(label_nulls <= 1, "at most one null in the batch");
+
+  RegSetIterator<Register> available_regs
+  // Z_R0 will be used to hold Z_R15(Z_SP) while pushing a new frame, So don't use that here.
+  // Z_R1 is will be used to hold r_bitmap in lookup_secondary_supers_table_var, so can't be used
+  // Z_R2, Z_R3, Z_R4 will be used in secondary_supers_verify, for the failure reporting
+    = (RegSet::range(Z_R0, Z_R15) - temps - sub_klass - super_klass - Z_R1_scratch - Z_R0_scratch - Z_R2 - Z_R3 - Z_R4).begin();
+
+  RegSet pushed_regs;
+
+  int slots = 0;
+  constexpr int NumRegs = 5;
+  Register* regs[NumRegs] = {&temp_reg, &temp2_reg, &temp3_reg, &temp4_reg, &result_reg};
```

Comment:
You don't need this array. It contains information that is in `pushed_regs`. Use `pushed_regs.begin()` to create an iterator, which you use like this:

```
    int i = 0;
    for (auto it = pushed_regs.begin(); *it != noreg; ++it, i++) {
     z_stg(*it++, i * BytesPerWord + frame::z_abi_160_size, Z_SP);
    }
```

Reply:
done.